pkg/sentry/server/validator/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,262 @@
+	sentryAudience string
+	controlPlaneNS string
+	controlPlaneTD spiffeid.TrustDomain
+}
+
+func New(ctx context.Context, opts Options) (validator.Validator, error) {
+	kubeClient, err := cl.NewForConfig(opts.RestConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cache, err := cache.New(opts.RestConfig, cache.Options{Scheme: scheme})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, obj := range []client.Object{
+		&metav1.PartialObjectMetadata{TypeMeta: metav1.TypeMeta{Kind: "Pod", APIVersion: "v1"}},
+		&configv1alpha1.Configuration{},
+	} {
+		if _, err := cache.GetInformer(ctx, obj); err != nil {
+			return nil, err
+		}
+	}
+
+	return &kubernetes{
+		auth:           kubeClient.AuthenticationV1(),
+		client:         cache,
+		ready:          cache.WaitForCacheSync,
+		sentryAudience: opts.SentryID.String(),
+		controlPlaneNS: opts.ControlPlaneNS,
+		controlPlaneTD: opts.SentryID.TrustDomain(),
+	}, nil
+}
+
+func (k *kubernetes) Start(ctx context.Context) error {
+	if err := k.client.(cache.Cache).Start(ctx); err != nil {
+		return err
+	}
+
+	<-ctx.Done()
+	return nil
+}
+
+func (k *kubernetes) Validate(ctx context.Context, req *sentryv1pb.SignCertificateRequest) (spiffeid.TrustDomain, error) {
+	if !k.ready(ctx) {
+		return spiffeid.TrustDomain{}, errors.New("validator not ready")
+	}
+
+	// The TrustDomain field is ignored by the Kubernetes validator.
+	if _, err := internal.Validate(ctx, req); err != nil {
+		return spiffeid.TrustDomain{}, err
+	}
+
+	prts, err := k.executeTokenReview(ctx, req.GetToken(), LegacyServiceAccountAudience, k.sentryAudience)
+	if err != nil {
+		return spiffeid.TrustDomain{}, err
+	}
+
+	if len(prts) != 4 || prts[0] != "system" {
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: provided token is not a properly structured service account token", errPrefix)
+	}
+
+	if prts[2] != req.Namespace {
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: namespace mismatch. received namespace: %s",
+			errPrefix, req.Namespace)
+	}
+
+	// We have already validated to the token against Kubernetes API server, so
+	// we do not need to supply a key.
+	ptoken, err := jwt.ParseInsecure([]byte(req.GetToken()), jwt.WithTypedClaim("kubernetes.io", new(k8sClaims)))
+	if err != nil {
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: failed to parse Kubernetes token: %s", errPrefix, err)
+	}
+	claimsT, ok := ptoken.Get("kubernetes.io")
+	if !ok {
+		return spiffeid.TrustDomain{}, errMissingPodClaim
+	}
+	claims, ok := claimsT.(*k8sClaims)
+	if !ok || len(claims.Pod.Name) == 0 {
+		return spiffeid.TrustDomain{}, errMissingPodClaim
+	}
+
+	var pod corev1.Pod
+	err = k.client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: claims.Pod.Name}, &pod)
+	if err != nil {
+		log.Errorf("failed to get pod %s/%s for requested identity: %s", req.Namespace, claims.Pod.Name, err)
+		return spiffeid.TrustDomain{}, fmt.Errorf("%s: failed to get pod of identity", errPrefix)
+	}
+	expID, ok := pod.GetAnnotations()[annotations.KeyAppID]
```

Comment:
Please don't read from the annotations, but inspect the flags passed to the `daprd` binary directly (also for the configurations below).

Although uncommon, it's entirely possible for someone to deploy an app on K8s skipping the Dapr Injector, by adding the `daprd` container directly. In this case, there'd be no annotations.

Reply:
This is wrong- whilst also fragile, the Kubernetes validator should be using Kubernetes primitives to evaluate the identity. Users who are not using the injector are still expected add appropriate annotations to pod templates.

Reply:
We should make sure this is clearly documented then. Right now I don't believe that's documented.